Clean up the use of package globals

a-palchikov opened the pull request:
This PR removes the need for the package globals for maintaining state - instead, all configuration (incl. from environment) is set on the root logger. The main reason for the change was the wish to set import environment externally instead of relying on magnet to invoke `make magnet-vars`.

The types have been split into two logical parts - the root logger (created with [Root](https://github.com/a-palchikov/magnet/blob/dmitri/buildx/magnet.go#L124)) and [Target](https://github.com/a-palchikov/magnet/blob/dmitri/buildx/magnet.go#L117)s which seems to be in line with the original intent.

I also removed the built-in mage target for dumping the environment  - this is best done in the client project. For this, the environment output API has been [repurposed](https://github.com/a-palchikov/magnet/blob/dmitri/buildx/common/common.go#L26) to accept the environment as arguments. Not sure if package `common` is the best place for it though.

@knisbet PTAL.

a-palchikov committed:
A few improvements

a-palchikov committed:
Do not set GOPATH as it breaks when set to the repository's root where go.mod exists

a-palchikov committed:
Fix linter warnings

a-palchikov committed:
Remove package globals. Store environment configuration in the root magnet. Split into root/child targets for easier separation.

a-palchikov committed:
Separate init/start of the status logger and delay init of secrets redactor

a-palchikov committed:
init status logger processes on-demand from a correspoding Target as it used to be

a-palchikov committed:
Make initOutput optional and do not block in Magnet.Shutdown

a-palchikov committed:
Remove unnecessary error return frmo Magnet.Shutdown

a-palchikov committed:
Update mod dependencies

a-palchikov committed:
Remove redundant BuildDir

a-palchikov committed:
Return magnet's environment with Env API

knisbet commented on logger.go:
```diff
@@ -83,11 +88,7 @@ func (s *SolveStatusLogger) tee() {
-					status.Logs[i].Data = []byte(strings.ReplaceAll(string(status.Logs[i].Data), value.Value, "<redacted>"))
-				}
-			}
+			status.Logs[i].Data = []byte(redactor.redact(string(status.Logs[i].Data)))
```
Probably should make redact just take []byte since this is the only use case instead of a bunch of casts here?

a-palchikov replied:
Yes, sure - will update.

knisbet reviewed (commented):
Overall I'm onboard, I think the only part I don't necessarily agree with is:

> I also removed the built-in mage target for dumping the environment - this is best done in the client project. For this, the environment output API has been repurposed to accept the environment as arguments. Not sure if package common is the best place for it though.

While it makes sense from a normal library development perspective, my thought with this although possibly naive, is to make common functionality easily importable. So something every project should have would just be an import for common. And then for any go project that wanted linting, could import the needed targets, although it might not easily be compatible with the dockerized builds.

This way across things like 10 projects there is less duplicate code, just one well written lint target that get's updated and then vendored / pulled into the repositories. That was the idea anyways with the approach, maybe it doesn't work in reality.

knisbet commented:
I think might also need to be a bit careful with the GOPATH stuff, as I remember that hackery was all done for builds that weren't converted to go modules yet... which might be necessary to support older versions of projects that are still shipping, unless there is a commitment to make sure everything is up to date.

a-palchikov commented:
> I think might also need to be a bit careful with the GOPATH stuff, as I remember that hackery was all done for builds that weren't converted to go modules yet... which might be necessary to support older versions of projects that are still shipping, unless there is a commitment to make sure everything is up to date.

Yeah, I'm in favor of simplifying things and assuming Go module support.
Let me know if there're active concerns with the shipped software (I don't know anything using this so far though).
I'll look into possible ways to provide this as configuration since currently gopath handling being done by the package only gets in the way.

knisbet commented:
Well I guess I'm not sure where these contributions are going /  what the intentions are. But the gravity main repo for 5.5/7.0 for M I believe is still on dep, unless the intention is to go through and bump to go modules for these older releases.

a-palchikov committed:
Allow overrides in GOPATH configuration handling in build containers

a-palchikov committed:
Update cache directory to be absolute in Config's checkAndSetDefaults to
minimize the use of os.Getwd.

a-palchikov commented:
> Well I guess I'm not sure where these contributions are going / what the intentions are. But the gravity main repo for 5.5/7.0 for M I believe is still on dep, unless the intention is to go through and bump to go modules for these older releases.

Sorry for not making myself completely clear. I was working on moving the tip (9.x) to your mage-based build and 5.5/7 branches are not currently in consideration. I did revert and rework the changes w.r.t GOPATH handling so I can configure the paths and there are defaults as before for the case if the paths are unspecified - PTAL.

Regarding the common functionality (which only included the dump of environment variables) - I agree that common functionality would make a lot of sense in general but since I moved the state in its entirety inside the root logger value (the type Magnet), there is no way to provide thi particular magefile target without global state anymore. I kept the output API so the equivalent target can be implemented easily in the package's client - let me know if that's something you don't agree with.

knisbet commented on golang.go:
```diff
@@ -592,3 +585,58 @@ func (m *GolangConfigTest) SetBuildContainer(value string) *GolangConfigTest {
+	// Defaults to process' working directory
+	hostPath string
+
+	// continerPath optionally specifies the path to the package repository inside
```
```suggestion
	// containerPath optionally specifies the path to the package repository inside
```

knisbet commented:
> > Well I guess I'm not sure where these contributions are going / what the intentions are. But the gravity main repo for 5.5/7.0 for M I believe is still on dep, unless the intention is to go through and bump to go modules for these older releases.
> 
> Sorry for not making myself completely clear. I was working on moving the tip (9.x) to your mage-based build and 5.5/7 branches are not currently in consideration. I did revert and rework the changes w.r.t GOPATH handling so I can configure the paths and there are defaults as before for the case if the paths are unspecified - PTAL.
> 
> Regarding the common functionality (which only included the dump of environment variables) - I agree that common functionality would make a lot of sense in general but since I moved the state in its entirety inside the root logger value (the type Magnet), there is no way to provide thi particular magefile target without global state anymore. I kept the output API so the equivalent target can be implemented easily in the package's client - let me know if that's something you don't agree with.

My suspicion is even moving tip to this, someone at some point will want it on older releases to test backports and the like. 

I agree so far the common functionality was just to dump env variables, but the intention was to try and see if it's possible to do alot more than that... import a common go target, and have targets for linting, testing, maybe coverage, security checks, etc. But it's also plausible that the targets themselves need to get duplicated, and it's all minimal library code to add that functionality and a bit of copy/pasta that doesn't change much between project and branch. 

I may be a little less worried about removing global state as well since it seems unrealistic to me that we'd want magnet to be initialized more than once independently of each other. But at the same time, this project is effectively dead besides you working on it, so I suppose I'm good as is.

a-palchikov committed:
Update shutdown hooks implementation to use a special variable instead

a-palchikov committed:
Remove Shutdown API in favor of Deinit special variable

knisbet commented:
I imagine the examples need to be updated to use Deinit as well.

a-palchikov committed:
Add support for coverage reporting for Go tests.

knisbet commented on golang.go:
```diff
@@ -443,25 +451,58 @@ func (m *GolangConfigTest) testDocker(ctx context.Context, packages ...string) e
+	}
+
+	return []string{
+		"bash", "-exc",
```
hmmm... do we prefer to build a pipeline in bash, or just exec / log the commands independently of eachother?

a-palchikov replied:
Please see 0fe1c91 for changes addressing this.

a-palchikov committed:
Update the behavior of dockerSrcPathFromGopath to use GOPATH as a prefix
when matching the host path. If filepath.Rel were used, it would always
match absolute path regardless of GOPATH setting.
Add stretchr/testify as a direct dependency.
Address review comments and fix linter warnings.

a-palchikov commented:
> I imagine the examples need to be updated to use Deinit as well.

Yes, true. I'll update the examples and the README.

a-palchikov committed:
Update and fix examples to sync with the latest changes.

a-palchikov commented:
> > I imagine the examples need to be updated to use Deinit as well.
> 
> Yes, true. I'll update the examples and the README.

Addressed in 5654ef6. @knisbet PTAL

a-palchikov committed:
Set ModulePath explicitly instead of adding go modules for examples

knisbet commented on dl.go:
```diff
@@ -24,13 +25,13 @@ type downloadMetadata struct {
 // Download begins a download of a url but doesn't block.
 // Returns a future that when called will block until it can return the path to the file on disk or an error.
-func (m *MagnetTarget) DownloadFuture(url string) func() (url string, path string, err error) {
+func (m *MagnetTarget) DownloadFuture(ctx context.Context, url string) func() (url string, path string, err error) {
```
Might want to consider adding a context to the future as well to make it return when cancelled, but not super important.

a-palchikov replied:
Addressed in https://github.com/gravitational/magnet/pull/2/commits/fa3aca08010b89e4bc5351a34cab37f248ff6fbc

knisbet replied:
This was my bad, the context to the base function should actually cancel the download, not sure what I was thinking.

knisbet commented on examples/golang/build.go:
```diff
@@ -74,15 +88,45 @@ func BuildContainer() (err error) {
+	defer func() { m.Complete(err) }()
+
+	return m.DockerBuild().
+		AddTag("builder:1.14").
```
maybe example-builer or something a little more clear the tag is just an example/test if looking at the image list sometime later?

a-palchikov replied:
Addressed in https://github.com/gravitational/magnet/pull/2/commits/fa3aca08010b89e4bc5351a34cab37f248ff6fbc

knisbet commented on examples/golang/build.go:
```diff
@@ -74,15 +88,45 @@ func BuildContainer() (err error) {
+	defer func() { m.Complete(err) }()
+
+	return m.DockerBuild().
+		AddTag("builder:1.14").
```
Also, instead of hardcode "builder:1.14" I think this would probably be a variable if not an example, so might be worth having the example the way we'd want it in a real project.

a-palchikov replied:
Addressed in https://github.com/gravitational/magnet/pull/2/commits/fa3aca08010b89e4bc5351a34cab37f248ff6fbc

a-palchikov committed:
Address review comments

knisbet commented on dl.go:
```diff
@@ -36,12 +36,16 @@ func (m *MagnetTarget) DownloadFuture(ctx context.Context, url string) func() (u
+				return url, "", trace.Wrap(err)
+			}
+			return url, path, nil
+		case <-ctx.Done():
```
This is my bad, not sure what I was thinking in the earlier comment.

The DownlodFuture takes a context itself and passes that to the underlying download function, does that not release properly if the context is cancelled? Otherwise it seems redundant that both the base function and the future each take a context that can cancel the download.

a-palchikov replied:
Yes, you're right. I should've thought about that as well.

knisbet commented on examples/golang/build.go:
```diff
@@ -27,14 +27,16 @@ import (
 	CacheDir:    "_build",
 	ModulePath:  "github.com/gravitational/magnet/examples/golang",
+	// If VERSION is unspecified, the default will be generated based on a git tag
+	Version: os.Getenv("VERSION"),
```
This seems a bit reversed to me, the magnet.Config should be the version specified by the magnet.EnvVar and not the os.Getenv. So I think the normal way to do this is this line becomes `Version: version,`
and the line60 `version = root.E...` will automatically retrieve the os.Getenv to override any of the envvars magnet supports: https://github.com/gravitational/magnet/blob/master/env.go#L35

a-palchikov replied:
Yes, this is a bit backwards but after I've removed the use of globals (which is this PR's primary goal), it would result in a circular dependency as `root` depends on `version` which depends on `root`. Maybe it would more "correct" to revert to the implicit globals in the magnet package for simplicity and instead remove the other bits that made use of the package cumbersome in some use-cases (e.g. implicit use of `make magnet-vars` to extract the otherwise custom configuration with no way of removing the reliance on the default `Makefile` file name, etc).

a-palchikov replied:
Addressed in 06fbac5

knisbet commented on examples/golang/build.go:
```diff
@@ -124,6 +135,10 @@ func Shutdown() {
 }
 
+func buildContainer() string {
+	return fmt.Sprint("build:", version)
```
nit: it was a problem in the original, but might be better to call this magnet-example-builder or something that looking at the docker image list is less likely to be confused by.

a-palchikov replied:
Addressed in 06fbac5

knisbet commented on examples/make_import/README.md:
```diff
@@ -24,8 +24,6 @@ Use the common help target:
-  DEBIAN_FRONTEND |       |             | Set to noninteractive or stderr to null to enable non-interactive output  
   GO_VERSION      |       | 1.9.2       | Set the golang version (Default from make)                                
   K8S_VERSION     |       | v1.6.6      | Set the k8s version (Default from make)                                   
-  XDG_CACHE_HOME  |       | build/cache | Location to store/cache build assets                         
```
Similar, this is a semi-standard way to specify a cache root, so is there a reason not to include this anymore?

a-palchikov replied:
One of the reasons was that this made the package more tied to a specific  implementation (freedesktop which is more of a linux thing).
If I revert the changes and move the environment variables back to the package globals this could be made a convenience again.

knisbet replied:
TBH I'm not sure if there is a better variable to standardize on that is consistent with mac/linux, or whether we even want the cache used for builds to share in that cache directory. But I do suspect there is value to letting it be configurable by the user in a standard way.

knisbet commented on examples/make_import/README.md:
```diff
@@ -24,8 +24,6 @@ Use the common help target:
         ENV       | VALUE |   DEFAULT   |                            SHORT DESCRIPTION                              
 ------------------+-------+-------------+---------------------------------------------------------------------------
   ARCH            |       | amd64       | Set the arch (Default from make)                                          
-  DEBIAN_FRONTEND |       |             | Set to noninteractive or stderr to null to enable non-interactive output  
```
did DEBIAN_FRONTEND get removed? I originally put this in to emulate the behaviour that many debian tools have that will pick up on this env var and change the output mode. I'm not aware of a generic env so this one seemed the most approrpiate.

a-palchikov replied:
Yes, I removed it but like I've been saying above I could reintroduce this and derive a default for it if I reverted the changes and moved environment back to package globals.
I know it's about configuration by convention more than anything - but I thought that having an [explicit configuration](https://github.com/gravitational/magnet/pull/2/files#diff-20346fa97ecbc96a69f300130936b45bdc4bfc57a73a204792178fb1d680067eR41) option would make the need for the implicit configuration less of an issue

knisbet replied:
Yea, the reason I hacked it up this way was to have configuration by convention, if you walk into any of these projects for some of the main interactions they work the same way. There are probably ways to approach it like move the vars into root and having a default that can optionally be called by the config when loaded or something. 

But I think having a consistent way to turn the interactive mode off will come up if/when this get's wired to a CI/CD type pipeline, to get usable logs on stdout.

knisbet replied:
I'm not opposed to making these MAGNET_ vars if that's the preferred idiom than trying to find something that makes sense in unix/mac lands, so `MAGNET_CACHE=...` / `MAGNET_OUTPUT_MODE=<noninteractive | interactive>`. Using a mode might make sense if we want to add more output formats in the future, like say json to feed to centralized logging.

a-palchikov replied:
Reverted the `DEBIAN_FRONTEND` and `XDG_CACHE_HOME` to keep the configuration-by-convention spirit and used them to infer the value for the progress flag. True, it is currently boolean but I'd keep it that way since this is as much as is needed from it (although buildkit does have it as a mode).

knisbet commented on examples/make_import/build.go:
```diff
@@ -55,3 +62,48 @@ func Env() (err error) {
+// importEnvFromMakefile invokes `make` to generate configuration for this mage script.
+// The script outputs a set of environment variables prefixed with `MAGNET_` which
+// are used as default values for the configuration variables defined by the script.
+func importEnvFromMakefile() (env map[string]string) {
```
All the parsing should be re-useable library code and not duplicated into each project.

a-palchikov replied:
Addressed in 84d297b

a-palchikov committed:
Address review comments

a-palchikov committed:
Address review comments

a-palchikov committed:
Revert the handling of environment variables to enable to define them
without relation to the builder's root target.
Revisit the docker container configuration for Go builds:
  * remove implicit host->container GOPATH handling
  * add the ability to override the GOPATH inside the container via
    configuration (not the environment) and compute container module
    path in GOPATH mode based on this configuration automatically.

a-palchikov committed:
Move environment handling to a type and allow overriding the external
configuration source via resetting the environment package global.

a-palchikov commented on common/common.go:
```diff
@@ -18,29 +18,25 @@ import (
 	"github.com/olekukonko/tablewriter"
 )
 
+// Help defines the utility namespace for help targets
```
I reverted the common utilities now that the environment handling is back with package globals.

a-palchikov commented on common/common.go:
```diff
@@ -18,29 +18,25 @@ import (
 	var result [][]string
 
-	for key, value := range magnet.EnvVars {
+	for key, value := range magnet.Env() {
```
`Env` merges the values (from imported source)

a-palchikov commented on docker.go:
```diff
@@ -382,9 +382,9 @@ func (m *DockerConfigRun) SetRemove(remove bool) *DockerConfigRun {
-// AddVolume attaches a filesystem mount to the container.
-func (m *DockerConfigRun) AddVolume(volume DockerBindMount) *DockerConfigRun {
-	m.Volumes = append(m.Volumes, volume)
+// AddVolume attaches a set of filesystem mounts to the container.
```
Added this while testing the docker builder changes and thought I'd leave it in.

a-palchikov commented on env.go:
```diff
@@ -19,64 +25,191 @@ type EnvVar struct {
-		if v, ok := ImportEnvVars[key]; ok {
-			return v
-		}
+// Environ represents the environment with configuration
```
Had to introduce an additional type for handling configuration from environment and allow the option to read overrides from an arbitrary source (e.g. in my case a custom makefile).

a-palchikov commented on env.go:
```diff
@@ -19,64 +25,191 @@ type EnvVar struct {
 	})
 }
+
+var debianFrontend = E(EnvVar{
```
Reintroduced the automatic configuration for progress using `DEBIAN_FRONTEND` unless explicitly configured and `XDG_CACHE_HOME` for cache location unless explicitly overridden with configuration.
The only downside now is that these I will not be able to infer from a custom configuration source but I'm fine with them being just in environment.

a-palchikov commented on examples/golang/Dockerfile:
```diff
@@ -4,11 +4,10 @@ ARG GOLANG_VER
 # Run a script to install golangci-lint
 RUN curl -sfL https://raw.githubusercontent.com/golangci/golangci-lint/master/install.sh| sh -s -- -b $GOPATH/bin ${GOLANGCI_VER}
 
-# install development libraries used when compiling fio
```
This was a left-over of copy-pasting I figured so removed it.

a-palchikov commented on golang.go:
```diff
@@ -130,27 +130,67 @@ type GolangConfigBuild struct {
+	// The value of the attribute will be written as `GOPATH` envar
+	// inside the container.
+	// Also sets `GO111MODULE=auto` envar if unspecified.
+	GOPath string
```
This is what I came up with instead of inferring the GOPATH from host. I think container has little to do with the host environment and the bit requiring the configuration is whether to use GOPATH and where it's at inside the container.

a-palchikov commented on magnet.go:
```diff
@@ -1,36 +1,65 @@
+
+	// PlainProgress specifies whether the logger uses fancy progress reporting.
+	// Set to true to see streaming output (e.g. on CI)
+	PlainProgress *bool
```
Keeping this as a flag since this currently suffices.

a-palchikov commented on magnet.go:
```diff
@@ -70,190 +86,220 @@ func (m *Magnet) printHeader() {
-func Root(c Config) *Magnet {
-	const statusChanSize = 128
+func Root(c Config) (*Magnet, error) {
+	if root != nil {
```
Instead of `sync.Once` and idempotency (concurrency is less of an issue, I think) the expected behavior would be to fail on subsequent initialization since only a single root target is expected at any time.

a-palchikov commented:
@knisbet PTAL. Addressed the rest of your comments in https://github.com/gravitational/magnet/pull/2/commits/9fff281e8070d3738fd1c0df38903d9616217bb9 and https://github.com/gravitational/magnet/pull/2/commits/92f830a2fe6c6b36baaf456424c96066f11d1c81

knisbet reviewed (approved):
We should probably merge at this point, nothing stops us from continuing to iterate as we live with it a bit and figure out what does and doesn't work well.

a-palchikov commented:
Can we merge (sorry I don't have merge permissions here)? Including the [mage dependency](https://github.com/knisbet/mage/pull/2) - I improved it slightly by removing the `Deinit` target from the list of targets since it was meant to be internal-use anyways.

knisbet commented:
> Can we merge (sorry I don't have merge permissions here)? Including the [mage dependency](https://github.com/knisbet/mage/pull/2) - I improved it slightly by removing the `Deinit` target from the list of targets since it was meant to be internal-use anyways.

My bad, I merged https://github.com/knisbet/mage/pull/2 since the fork is on my personal repository, and added external contributors to this repository. @a-palchikov you probably should be able to merge now.

a-palchikov committed:
Bump mage dependency to master
